R/project.R

```diff
@@ -253,59 +254,75 @@ project <- function(x, R, si, n_sim = 100, n_days = 7,
   ## vector for R of length `n + 1`, we assume the `R` is constant per time
   ## period and input is silently converted to a list of length `n + 1`.
 
-  time_period <- 1L
-  
+
   if (!is.list(R)) { # i.e. R is a vector and needs conversion to a list
     if (n_time_periods > 1L) {  # several time periods
       R <- as.list(R)
-     } else {  # a single time period
-       R <- list(R)
-     }
+    } else {  # a single time period
+      R <- list(R)
+    }
   }
-  
+
 
   ## On the drawing of R values: either these are constant within simulations,
   ## so drawn once for all simulations, or they need drawing at every time step
   ## for every simulations.
-
+  
   ## On the handling of reporting: reporting first affects the force of
   ## infection lambda, with the underlying assumption that the true epicurve
   ## multiplied by the (constant) reporting results in the observed one. Then,
   ## the true incidence (true_I) is determined using this lambda (Poisson
   ## process) and we apply effects sampling to this true incidence to get the
   ## projected one, using a Binomial sampling.
 
+  if (is.finite(time_change)) {
+    time_change_boundaries <- c(1, time_change, t_stop+1)
+  } else {
+    time_change_boundaries <- c(1, t_stop+1)
+  }
+  R_t <- matrix(nrow = 0, ncol = n_sim)
   if (R_fix_within) {
-    R_time_period <- R[[time_period]]
-    current_R <- sample_(R_time_period, n_sim, replace = TRUE)
+    for (time_period in 1:n_time_periods) {
+      R_time_period <- sample_(R[[time_period]], n_sim, replace = TRUE)
+      period_duration <- time_change_boundaries[time_period+1] - time_change_boundaries[time_period]
+      current_R_t <- do.call(
+        'rbind', 
+        replicate(period_duration, R_time_period, simplify = FALSE)
+      )
+      R_t <- rbind(R_t, current_R_t)
+    }
+  } else {
+    time_period <- 1L
+    for (i in 1:t_stop) {
+      R_time_period <- R[[time_period]]
+      current_R_t <- sample_(R_time_period, n_sim, replace = TRUE)
+      R_t <- rbind(R_t, current_R_t)
+      if (i %in% time_change) {
+        time_period <- time_period + 1
+      }
+    }
   }
+  rownames(R_t) <- NULL
 
   for (i in t_sim) {
-    ## update the time period if needed
-    if (i %in% time_change) {
-      time_period <- time_period + 1
-    }
-     
-    if (!R_fix_within) {
-    R_time_period <- R[[time_period]]
-      current_R <- sample_(R_time_period, n_sim, replace = TRUE)
-    }
-    lambda <- utils::tail(ws, i) %*% out[seq_len(i), , drop = FALSE]
+
+    lambda_R <- ws[(length(ws)-(i-1)):(length(ws)-1)] %*%
```

Comment:
I think the reversing here may be wrong, as `ws` is already the reverse of the pmf of the serial interval.

Reply:
I think this is correct in both versions, although maybe writing something like `si[(i-1):1]` and not defining a `ws` might be better understandable, but surely it's a matter of taste... `ws` is not reversed, just taking the last elements (the last one excluded in the new version).